File: core/src/com/megaman/game/entities/impl/projectiles/ProjectileFactory.java
package com.megaman.game.entities.impl.projectiles;

import com.badlogic.gdx.utils.ObjectMap;
import com.megaman.game.MegamanGame;
import com.megaman.game.entities.Entity;
import com.megaman.game.entities.utils.factories.EntityFactory;
import com.megaman.game.entities.utils.factories.EntityPool;
import com.megaman.game.entities.impl.projectiles.impl.*;

public class ProjectileFactory implements EntityFactory {

    public static final String BULLET = "Bullet";
    public static final String PICKET = "Picket";
    public static final String JOEBALL = "Joeball";
    public static final String FIREBALL = "Fireball";
    public static final String SNOWBALL = "Snowball";
    public static final String CHARGED_SHOT = "ChargedShot";
    public static final String PRECIOUS_SHOT = "PreciousShot";

    private final ObjectMap<String, EntityPool> pools;

    public ProjectileFactory(MegamanGame game) {
        this.pools = new ObjectMap<>() {{
            put(BULLET, new EntityPool(20, () -> new Bullet(game)));
            put(PICKET, new EntityPool(10, () -> new Picket(game)));
            put(JOEBALL, new EntityPool(3, () -> new JoeBall(game)));
            put(FIREBALL, new EntityPool(3, () -> new Fireball(game)));
            put(SNOWBALL, new EntityPool(3, () -> new Snowball(game)));
            put(CHARGED_SHOT, new EntityPool(2, () -> new ChargedShot(game)));
            put(PRECIOUS_SHOT, new EntityPool(10, () -> new PreciousShot(game)));
        }};
    }

    @Override
    public Entity fetch(String key) {
        return pools.get(key).fetch();
    }

}

File: core/src/com/megaman/maverick/game/entities/factories/impl/ProjectileFactory.kt
package com.megaman.maverick.game.entities.factories.impl

import com.badlogic.gdx.utils.ObjectMap
import com.engine.common.GameLogger
import com.engine.common.objects.Pool
import com.engine.entities.IGameEntity
import com.engine.factories.IFactory
import com.megaman.maverick.game.MegamanMaverickGame
import com.megaman.maverick.game.entities.IProjectileEntity
import com.megaman.maverick.game.entities.factories.EntityPoolCreator
import com.megaman.maverick.game.entities.projectiles.Bullet
import com.megaman.maverick.game.entities.projectiles.ChargedShot

/** A factory that creates projectiles. */
class ProjectileFactory(game: MegamanMaverickGame) : IFactory<IGameEntity> {

  companion object {
    const val TAG = "ProjectileFactory"

    const val BULLET = "Bullet"
    const val PICKET = "Picket"
    const val JOEBALL = "Joeball"
    const val FIREBALL = "Fireball"
    const val SNOWBALL = "Snowball"
    const val CHARGED_SHOT = "ChargedShot"
    const val PRECIOUS_SHOT = "PreciousShot"
  }

  private val pools = ObjectMap<Any, Pool<IProjectileEntity>>()

  init {
    // bullet
    pools.put(BULLET, EntityPoolCreator.create(10) { Bullet(game) })
    // charged shot
    pools.put(CHARGED_SHOT, EntityPoolCreator.create(5) { ChargedShot(game) })
  }

  /**
   * Fetches a projectile from the pool.
   *
   * @param key The key of the projectile to fetch.
   * @return A projectile from the pool.
   */
  override fun fetch(key: Any): IProjectileEntity? {
    GameLogger.debug(TAG, "Spawning Projectile: key = $key")
    return pools.get(if (key == "") BULLET else key)?.fetch()
  }
}
